theano/gof/destroyhandler.py

```diff
@@ -394,6 +394,32 @@ def get_destroyers_of(r):
                 return []
         fgraph.destroyers = get_destroyers_of
 
+        def has_destroyers(protected_list):
```

Comment:
Are you sure that ```self.stale_droot``` is a good condition? I quickly implemented this : 

```python
Supervisor:
def validate(self, fgraph):
        if not hasattr(fgraph, 'has_destroyers'):
            return True
        if fgraph.has_destroyers(self.protected):
            raise gof.InconsistencyError("Trying to destroy a protected"
                                         "Variable.")
DestroyHandler:
def unpickle(self, fgraph):
        def get_destroyers_of(r):
            droot, _, root_destroyer = self.refresh_droot_impact()
            try:
                return [root_destroyer[droot[r]]]
            except Exception:
                return []
        fgraph.destroyers = get_destroyers_of

        def recursive_destroys_finder(protected_var):
            # protected_var is the idx'th input of app.
            for (app, idx) in protected_var.clients:
                if app == 'output':
                    continue
                destroy_maps = getattr(app.op, 'destroy_map', {}).values()
                # If True means that the apply node, destroys the protected_var.
                if idx in [dmap for sublist in destroy_maps for dmap in sublist]:
                    return True
                for var_idx in getattr(app.op, 'view_map', {}).keys():
                    if idx in app.op.view_map[var_idx]:
                        # We need to recursivly check the destroy_map of all the
                        # outputs that we have a view_map on.
                        if recursive_destroys_finder(app.outputs[var_idx]):
                            return True
            return False

        def has_destroyers(protected_list):
            if not self.stale_droot:
                for var in protected_list + list(fgraph.outputs):
                    if get_destroyers_of(var):
                        return True
                return False

            for protected_var in protected_list:
                if recursive_destroys_finder(protected_var):
                    return True
            return False

        fgraph.has_destroyers = has_destroyers
```

But I had many errors with the OutputGuard:
```
NotImplementedError: ('OutputGuard.0 is a view/destroyed version of more then one inputs. Currently, we only support the case where an output is a view or a destroyed version of one input.', 'Test <function <lambda> at 0x7fac379dbb18>::x2: Error occurred while computing the gradient on the following inputs: [array([[-0.07363263, -0.66992872,  0.980925  ]])]')
```

 So I guess the condition is not correct. I think the could would be nicer with your change but why do you think this will be a speedup?